packages/opentelemetry-context-async-hooks/src/AsyncHooksContextManager.ts

```diff
@@ -72,7 +89,34 @@ export class AsyncHooksContextManager implements ContextManager {
       if (oldContext === undefined) {
         this._destroy(uid);
       } else {
-        this._contexts[uid] = oldContext;
+        ref.set(oldContext);
+      }
+    }
+  }
+
+  async withAsync<T extends Promise<any>, U extends (...args: unknown[]) => T>(
```

Comment:
the body of those 2 functions is almost identical, except await.
I wonder if that make sense to combine them by simply exporting 2 functionalities for example
1. `setContext` or `saveContext` ?
2. `restoreContext`

for example
```
async withAsync<T extends Promise<any>, U extends (...args: unknown[]) => T>(
    context: Context,
    fn: U
  ): Promise<T> {
    const {uid, oldContext} = setContext(context);
    try {
      return await fn();
    } catch (err) {
      throw err;
    } finally {
      restoreContext(uid, oldContext);
    }
  }
  
with<T extends (...args: unknown[]) => ReturnType<T>>(
    context: Context,
    fn: T
  ): ReturnType<T> {
    const {uid, oldContext} = setContext(context);
    try {
      return fn();
    } catch (err) {
      throw err;
    } finally {
      restoreContext(uid, oldContext);
    }
  }  
```
the reason is that it will be easier to change things and probably less errors, WDYT ?

Reply:
I agree that they are pretty much the same, however i believe `with()` doesn't need  to store reference like i've done so i might end up re-write that part to avoid the memory overhead  (since `with`) is the most used. I would prefer to do more test like this and refactor later, WDYT ?

Reply:
I would definitely prefer to keep them separate for now. Until we know that `withAsync` doesn't fail on some common use case or something. Maybe it might be worth replacing all instances of `with` with `withAsync` locally and running the tests to see if anything breaks?

Reply:
The only difference being the `await`, it will just work since v8 will ignore it i believe

Reply:
Maybe we should run the benchmarks with both and see how big the difference really is. My worry is that async contexts are created very often so any small overhead could turn out disproportionately large impact.

Reply:
Do you want it done in this specific PR or in a seperate one ?

Reply:
but this is just code duplication for now, isn't it ?
`however i believe with() doesn't need to store reference like i've done`
but now they are looking the same, there is no difference with either setting or restoring context - the code for both is identical or am I missing something ?

Reply:
It is, is this a blocker for you ?

Reply:
no , I have already approved :)

Reply:
I would prefer to have a single function, but I'm concerned what the performance fallout would be if we made all `with` calls async. 

@vmarchaud if possible I would like the `with` to be able to not have the overhead that was introduced with the new reference object

Reply:
> I would prefer to have a single function, but I'm concerned what the performance fallout would be if we made all with calls async.

Even if there were any performance fallout of having one implementation now, it may in the future since v8 optimization change over time. I also like the fact that it forces developers to think about which case they are in (even if its harder).

@dyladan I will try to do that tomorrow then

Reply:
I think we are talking about 2 different things in the same thread. I was just pointing that code duplication might be simply reduce to make it easier for testing and for readiness as well. Not for merging or making just one function which is async by default. The async function are significantly slower then no async. But having the same piece of code in both and simply refactoring this to use function with name what it really happening there is all what I was trying to say here. But a big no for having a default one async `with` function

Reply:
@obecny Yeah sorry for the mix up, indeed we are still speaking about having two different functions, one sync and one async.